sdk/core/core/src/az_url.c

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "az_hex_private.h"
+#include <az_precondition_internal.h>
+#include <az_url_internal.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD AZ_INLINE bool _az_url_should_encode(uint8_t c)
+{
+  switch (c)
+  {
+    case '-':
+    case '_':
+    case '.':
+    case '~':
+      return false;
+    default:
+      return !(('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'));
+  }
+}
+
+AZ_NODISCARD az_result _az_url_encode(az_span destination, az_span source, int32_t* out_length)
+{
+  AZ_PRECONDITION_NOT_NULL(out_length);
+  AZ_PRECONDITION_VALID_SPAN(source, 0, true);
```

Comment:
```suggestion
  AZ_PRECONDITION_VALID_SPAN(source, 0, false);
```

Reply:
No, I mean that null is fine, if size is 0.
Because why do I treat `az_span empty = { 0 }` and `az_span empty = AZ_SPAN_FROM_STR("")` differently?
but actually, I am working on the proposal PR to remove is_null_valid parameter all together.

Reply:
You don't want accidental null pointer access (or dereference).

This method is grabbing the az_span_ptr and I don't think we should allow null ptrs to avoid accidental bugs.

Reply:
Dereferencing any pointer without checkin the size first is an error already, regardless of whether it is pointer to null, orpointer to garbage. The size is 0, you should not be accessing the data that is out of bounds.
This code does not dereference the ponter, it checks the size, and only accesses the indices that are in range.

Reply:
Per our online conversation, I'll update preconditions to allow NULL-ptr input spans (where applicable), and disallow output spans to have NULLs, even if input size is 0.

Reply:
You can see example in this function, I'll apply this principle in more places in our code.